File: desktop/packages/desktop/src/window.ts
/**
 * Window management API for Neutron Desktop.
 */

declare const __TAURI__: {
  core: {
    invoke: (cmd: string, args?: Record<string, unknown>) => Promise<unknown>;
  };
};

export interface WindowOptions {
  title?: string;
  width?: number;
  height?: number;
  x?: number;
  y?: number;
  resizable?: boolean;
  decorations?: boolean;
  transparent?: boolean;
  fullscreen?: boolean;
}

export class NeutronWindow {
  constructor(public readonly label: string) {}

  async setTitle(title: string): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|set_title', { label: this.label, title });
  }

  async setSize(width: number, height: number): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|set_size', { label: this.label, width, height });
  }

  async setPosition(x: number, y: number): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|set_position', { label: this.label, x, y });
  }

  async center(): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|center', { label: this.label });
  }

  async minimize(): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|minimize', { label: this.label });
  }

  async maximize(): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|maximize', { label: this.label });
  }

  async unmaximize(): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|unmaximize', { label: this.label });
  }

  async toggleMaximize(): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|toggle_maximize', { label: this.label });
  }

  async setFullscreen(fullscreen: boolean): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|set_fullscreen', { label: this.label, fullscreen });
  }

  async close(): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|close', { label: this.label });
  }

  async hide(): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|hide', { label: this.label });
  }

  async show(): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|show', { label: this.label });
  }

  async focus(): Promise<void> {
    await __TAURI__.core.invoke('plugin:window|focus', { label: this.label });
  }
}

/** Get the current (main) window. */
export function getCurrentWindow(): NeutronWindow {
  return new NeutronWindow('main');
}

/** Create a new window. */
export async function createWindow(label: string, options?: WindowOptions): Promise<NeutronWindow> {
  await __TAURI__.core.invoke('plugin:window|create', { label, ...options });
  return new NeutronWindow(label);
}

File: desktop/crates/neutron-desktop/src/window.rs
/// Configuration for the main application window.
pub struct WindowConfig {
    pub title: String,
    pub width: f64,
    pub height: f64,
    pub resizable: bool,
    pub decorations: bool,
    pub transparent: bool,
    pub fullscreen: bool,
    pub min_width: Option<f64>,
    pub min_height: Option<f64>,
}

impl WindowConfig {
    pub fn title(&mut self, title: impl Into<String>) -> &mut Self {
        self.title = title.into();
        self
    }

    pub fn size(&mut self, width: f64, height: f64) -> &mut Self {
        self.width = width;
        self.height = height;
        self
    }

    pub fn min_size(&mut self, width: f64, height: f64) -> &mut Self {
        self.min_width = Some(width);
        self.min_height = Some(height);
        self
    }

    pub fn resizable(&mut self, resizable: bool) -> &mut Self {
        self.resizable = resizable;
        self
    }

    pub fn decorations(&mut self, decorations: bool) -> &mut Self {
        self.decorations = decorations;
        self
    }

    pub fn transparent(&mut self, transparent: bool) -> &mut Self {
        self.transparent = transparent;
        self
    }

    pub fn fullscreen(&mut self, fullscreen: bool) -> &mut Self {
        self.fullscreen = fullscreen;
        self
    }
}

impl Default for WindowConfig {
    fn default() -> Self {
        Self {
            title: "Neutron App".to_string(),
            width: 1200.0,
            height: 800.0,
            resizable: true,
            decorations: true,
            transparent: false,
            fullscreen: false,
            min_width: None,
            min_height: None,
        }
    }
}
